service/history/replication/ack_manager.go

```diff
@@ -596,6 +589,72 @@ func (p *ackMgrImpl) processReplication(
 	}
 }
 
+func (p *ackMgrImpl) processNewRunReplication(
+	ctx context.Context,
+	namespaceID namespace.ID,
+	workflowID string,
+	newRunID string,
+	branchToken []byte,
+	taskVersion int64,
+	task *replicationspb.ReplicationTask,
+) (retReplicationTask *replicationspb.ReplicationTask, retError error) {
+
+	attr, ok := task.Attributes.(*replicationspb.ReplicationTask_HistoryTaskAttributes)
+	if !ok {
+		return nil, serviceerror.NewInternal("Wrong replication task to process new run replication.")
+	}
+
+	var newRunBranchToken []byte
+	if len(newRunID) > 0 {
+		newRunContext, releaseFn, err := p.workflowCache.GetOrCreateWorkflowExecution(
+			ctx,
+			namespaceID,
+			commonpb.WorkflowExecution{
+				WorkflowId: workflowID,
+				RunId:      newRunID,
+			},
+			workflow.CallerTypeTask,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		newRunMutableState, err := newRunContext.LoadMutableState(ctx)
+		if err != nil {
+			releaseFn(err)
+			return nil, err
+		}
+		_, newRunBranchToken, err = getVersionHistoryItems(
+			newRunMutableState,
+			common.FirstEventID,
+			taskVersion,
+		)
+		if err != nil {
+			releaseFn(err)
```

Comment:
if the intention is to reduce locking, then maybe create a function just for getting the branch token? then reuse this function for normal runs & continue as new runs